File: sample/src/main/java/de/rheinfabrik/heimdalldroid/network/oauth2/TraktTvRefreshAccessTokenGrant.java
package de.rheinfabrik.heimdalldroid.network.oauth2;

import de.rheinfabrik.heimdall.OAuth2AccessToken;
import de.rheinfabrik.heimdall.grants.OAuth2RefreshAccessTokenGrant;
import de.rheinfabrik.heimdalldroid.network.TraktTvApiFactory;
import de.rheinfabrik.heimdalldroid.network.models.RefreshTokenRequestBody;
import rx.Observable;

/**
 * TraktTv refresh token grant as described in http://docs.trakt.apiary.io/#reference/authentication-oauth/token/exchange-refresh_token-for-access_token.
 */
public class TraktTvRefreshAccessTokenGrant extends OAuth2RefreshAccessTokenGrant<OAuth2AccessToken> {

    // Properties

    public String clientSecret;
    public String clientId;
    public String redirectUri;

    // OAuth2RefreshAccessTokenGrant

    @Override
    public Observable<OAuth2AccessToken> grantNewAccessToken() {
        RefreshTokenRequestBody body = new RefreshTokenRequestBody(refreshToken, clientId, clientSecret, redirectUri, GRANT_TYPE);
        return TraktTvApiFactory.newApiService().refreshAccessToken(body);
    }
}

File: sample/src/main/java/de/rheinfabrik/heimdalldroid/network/oauth2/coroutines/TraktTvRefreshAccessTokenGrant.kt
package de.rheinfabrik.heimdalldroid.network.oauth2.coroutines

import de.rheinfabrik.heimdall2.accesstoken.OAuth2AccessToken
import de.rheinfabrik.heimdall2.coroutines.grants.OAuth2RefreshAccessTokenGrant
import de.rheinfabrik.heimdalldroid.network.TraktTvApiFactory
import de.rheinfabrik.heimdalldroid.network.models.RefreshTokenRequestBody

class TraktTvRefreshAccessTokenGrant(
    val clientSecret: String,
    val clientId: String,
    val redirectUri: String
) : OAuth2RefreshAccessTokenGrant() {

    // OAuth2RefreshAccessTokenGrant
    override suspend fun grantNewAccessToken(): OAuth2AccessToken =
        TraktTvApiFactory.newApiServiceCoroutines().refreshAccessToken(
            body = RefreshTokenRequestBody(
                refreshToken,
                clientId,
                clientSecret,
                redirectUri,
                GRANT_TYPE
            )
        )
}
